File: app/src/main/java/com/mkulesh/onpc/iscp/messages/DcpAudioRestorerMsg.java
/*
 * Enhanced Music Controller
 * Copyright (C) 2018-2023 by Mikhail Kulesh
 *
 * This program is free software: you can redistribute it and/or modify it under the terms of the GNU
 * General Public License as published by the Free Software Foundation, either version 3 of the License,
 * or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without
 * even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details. You should have received a copy of the GNU General
 * Public License along with this program.
 */

package com.mkulesh.onpc.iscp.messages;

import com.mkulesh.onpc.R;
import com.mkulesh.onpc.iscp.EISCPMessage;
import com.mkulesh.onpc.iscp.ISCPMessage;

import java.util.ArrayList;
import java.util.Collections;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.StringRes;

/*
 * Denon control protocol - DCP audio restorer
 */
public class DcpAudioRestorerMsg extends ISCPMessage
{
    public final static String CODE = "D04";
    private final static String DCP_COMMAND = "PSRSTR";

    @NonNull
    public static ArrayList<String> getAcceptedDcpCodes()
    {
        return new ArrayList<>(Collections.singletonList(DCP_COMMAND));
    }

    public enum Status implements DcpStringParameterIf
    {
        NONE("N/A", R.string.device_dcp_audio_restorer_none),
        OFF("OFF", R.string.device_dcp_audio_restorer_off),
        LOW("LOW", R.string.device_dcp_audio_restorer_low),
        MED("MED", R.string.device_dcp_audio_restorer_medium),
        HI("HI", R.string.device_dcp_audio_restorer_high);

        final String code;

        @StringRes
        final int descriptionId;

        Status(final String code, @StringRes final int descriptionId)
        {
            this.code = code;
            this.descriptionId = descriptionId;
        }

        public String getCode()
        {
            return code;
        }

        @NonNull
        public String getDcpCode()
        {
            return code;
        }

        @StringRes
        public int getDescriptionId()
        {
            return descriptionId;
        }
    }

    private final Status status;

    DcpAudioRestorerMsg(EISCPMessage raw) throws Exception
    {
        super(raw);
        status = (Status) searchParameter(data, Status.values(), Status.NONE);
    }

    public DcpAudioRestorerMsg(Status status)
    {
        super(0, null);
        this.status = status;
    }

    public Status getStatus()
    {
        return status;
    }

    @NonNull
    @Override
    public String toString()
    {
        return CODE + "[" + status.toString() + "]";
    }

    @Override
    public EISCPMessage getCmdMsg()
    {
        return new EISCPMessage(CODE, status.getCode());
    }

    @Override
    public boolean hasImpactOnMediaList()
    {
        return false;
    }

    @Nullable
    public static DcpAudioRestorerMsg processDcpMessage(@NonNull String dcpMsg)
    {
        final Status s = (Status) searchDcpParameter(DCP_COMMAND, dcpMsg, Status.values());
        return s != null ? new DcpAudioRestorerMsg(s) : null;
    }

    @Nullable
    @Override
    public String buildDcpMsg(boolean isQuery)
    {
        // A space is needed for this command
        return DCP_COMMAND + " " + (isQuery ? DCP_MSG_REQ : status.getDcpCode());
    }

    public static Status toggle(Status s)
    {
        switch (s)
        {
        case OFF:
            return Status.LOW;
        case LOW:
            return Status.MED;
        case MED:
            return Status.HI;
        case HI:
            return Status.OFF;
        default:
            return Status.NONE;
        }
    }
}

File: lib/iscp/messages/DcpAudioRestorerMsg.dart
/*
 * Enhanced Music Controller
 * Copyright (C) 2018-2023 by Mikhail Kulesh
 *
 * This program is free software: you can redistribute it and/or modify it under the terms of the GNU
 * General Public License as published by the Free Software Foundation, either version 3 of the License,
 * or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without
 * even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details. You should have received a copy of the GNU General
 * Public License along with this program.
 */
// @dart=2.9

import "../../constants/Strings.dart";
import "../EISCPMessage.dart";
import "EnumParameterMsg.dart";

enum DcpAudioRestorer
{
    NONE,
    OFF,
    LOW,
    MED,
    HI
}

/*
 * Denon control protocol - DCP audio restorer
 */
class DcpAudioRestorerMsg extends EnumParameterMsg<DcpAudioRestorer>
{
    static const String CODE = "D04";
    static const String _DCP_COMMAND = "PSRSTR";

    static List<String> getAcceptedDcpCodes()
    => [ _DCP_COMMAND ];

    static const ExtEnum<DcpAudioRestorer> ValueEnum = ExtEnum<DcpAudioRestorer>([
        EnumItem.code(DcpAudioRestorer.NONE, "N/A",
            descrList: Strings.l_device_dcp_audio_restorer_none, defValue: true),
        EnumItem.code(DcpAudioRestorer.OFF, "OFF",
            descrList: Strings.l_device_dcp_audio_restorer_off),
        EnumItem.code(DcpAudioRestorer.LOW, "LOW",
            descrList: Strings.l_device_dcp_audio_restorer_low),
        EnumItem.code(DcpAudioRestorer.MED, "MED",
            descrList: Strings.l_device_dcp_audio_restorer_medium),
        EnumItem.code(DcpAudioRestorer.HI, "HI",
            descrList: Strings.l_device_dcp_audio_restorer_high)
    ]);

    DcpAudioRestorerMsg(EISCPMessage raw) : super(CODE, raw, ValueEnum);

    DcpAudioRestorerMsg.toggle(EnumItem<DcpAudioRestorer> s) :
            super.output(CODE, _toggle(s.key), ValueEnum);

    DcpAudioRestorerMsg._dcp(DcpAudioRestorer v) : super.output(CODE, v, ValueEnum);

    @override
    bool hasImpactOnMediaList()
    => false;

    static DcpAudioRestorerMsg processDcpMessage(String dcpMsg)
    {
        final EnumItem<DcpAudioRestorer> s = ValueEnum.valueByDcpCommand(_DCP_COMMAND, dcpMsg);
        return s != null ? DcpAudioRestorerMsg._dcp(s.key) : null;
    }

    @override
    String buildDcpMsg(bool isQuery)
    => buildDcpRequest(isQuery, _DCP_COMMAND, sep: " ");

    static DcpAudioRestorer _toggle(DcpAudioRestorer s)
    {
        switch (s)
        {
        case DcpAudioRestorer.OFF:
            return DcpAudioRestorer.LOW;
        case DcpAudioRestorer.LOW:
            return DcpAudioRestorer.MED;
        case DcpAudioRestorer.MED:
            return DcpAudioRestorer.HI;
        case DcpAudioRestorer.HI:
            return DcpAudioRestorer.OFF;
        default:
            return DcpAudioRestorer.NONE;
        }
    }
}